File: src/Application Specific/Classes/Entity Collections/Asset/AssetSQL.vb
Imports System.Data.SqlClient

Namespace Entity

    Namespace Assets

        Public Class AssetSQL
            Private _database As Sys.Database

            Public Sub New(ByVal database As Sys.Database)
                _database = database
            End Sub

#Region "Functions"

            Public Sub Delete(ByVal AssetID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@AssetID", AssetID, True)
                _database.ExecuteSP_NO_Return("Asset_Delete")
            End Sub

            Public Function [Asset_Get](ByVal AssetID As Integer) As Asset
                _database.ClearParameter()
                _database.AddParameter("@AssetID", AssetID)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Asset_Get")

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oAsset As New Asset
                        With oAsset
                            .IgnoreExceptionsOnSetMethods = True
                            .SetAssetID = dt.Rows(0).Item("AssetID")
                            .SetPropertyID = dt.Rows(0).Item("SiteID")
                            .SetGasTypeID = Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("GasTypeID"))
                            .SetFlueTypeID = Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("FlueTypeID"))
                            .SetProductID = dt.Rows(0).Item("ProductID")
                            .SetSerialNum = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("SerialNum"))
                            If IsDBNull(dt.Rows(0).Item("DateFitted")) Then
                                .DateFitted = Date.MinValue
                            Else
                                .DateFitted = CDate(dt.Rows(0).Item("DateFitted"))
                            End If
                            If IsDBNull(dt.Rows(0).Item("CertificateLastIssued")) Then
                                .CertificateLastIssued = Date.MinValue
                            Else
                                .CertificateLastIssued = CDate(dt.Rows(0).Item("CertificateLastIssued"))
                            End If

                            If IsDBNull(dt.Rows(0).Item("LastServicedDate")) Then
                                .LastServicedDate = Date.MinValue
                            Else
                                .LastServicedDate = CDate(dt.Rows(0).Item("LastServicedDate"))
                            End If

                            If IsDBNull(dt.Rows(0).Item("WarrantyStartDate")) Then
                                .WarrantyStartDate = Date.MinValue
                            Else
                                .WarrantyStartDate = CDate(dt.Rows(0).Item("WarrantyStartDate"))
                            End If

                            If IsDBNull(dt.Rows(0).Item("WarrantyEndDate")) Then
                                .WarrantyEndDate = Date.MinValue
                            Else
                                .WarrantyEndDate = CDate(dt.Rows(0).Item("WarrantyEndDate"))
                            End If

                            .SetBoughtFrom = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("BoughtFrom"))
                            .SetSuppliedBy = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("SuppliedBy"))
                            .SetLocation = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("Location"))
                            .SetNotes = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("Notes"))
                            .SetGCNumber = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("GCNumber"))
                            .SetYearOfManufacture = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("YearOfManufacture"))
                            .SetApproximateValue = Entity.Sys.Helper.MakeDoubleValid(dt.Rows(0).Item("ApproximateValue"))
                            .SetDeleted = dt.Rows(0).Item("Deleted")
                            .SetActive = Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("Active"))
                            .SetTenantsAppliance = Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("TenantsAppliance"))
                        End With
                        oAsset.Exists = True
                        Return oAsset
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function Check_Unique_Serial(ByVal Serial As String, ByVal ID As Integer) As Boolean
                Dim NumberOfAssets As Integer = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteScalar("SELECT COUNT(AssetID) as 'NumberOfAssets' " &
                "FROM tblasset WHERE SerialNum = '" & Serial & "' AND AssetID <> " & ID, False))

                If NumberOfAssets = 0 Then
                    Return True
                Else
                    Return False
                End If
            End Function

            Public Function [Asset_GetAll](ByVal userID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@UserID", userID)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Asset_GetAll_Mk1")
                dt.TableName = Entity.Sys.Enums.TableNames.tblAsset.ToString
                Return New DataView(dt)
            End Function

            Public Function [Asset_GetForCustomer](ByVal CustomerID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@CustomerID", CustomerID, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("[Asset_GetForCustomer]")
                dt.TableName = Entity.Sys.Enums.TableNames.tblAsset.ToString
                Return New DataView(dt)
            End Function

            Public Function [Asset_GetForSite](ByVal SiteID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@SiteID", SiteID, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("[Asset_GetForSite]")
                dt.TableName = Entity.Sys.Enums.TableNames.tblAsset.ToString
                Return New DataView(dt)
            End Function

            Public Function [Asset_Search](ByVal Criteria As String, Optional ByVal RegionID As Integer = 0) As DataView
                _database.ClearParameter()
                _database.AddParameter("@SearchString", Criteria, True)
                _database.AddParameter("@RegionID", RegionID, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("[Asset_SearchList]")
                dt.TableName = Entity.Sys.Enums.TableNames.tblAsset.ToString
                Return New DataView(dt)
            End Function

            Public Function [Insert](ByVal oAsset As Asset) As Asset
                _database.ClearParameter()
                AddAssetParametersToCommand(oAsset)

                oAsset.SetAssetID = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Asset_Insert"))
                oAsset.Exists = True
                Return oAsset
            End Function

            Public Sub [Update](ByVal oAsset As Asset)
                _database.ClearParameter()
                _database.AddParameter("@AssetID", oAsset.AssetID, True)
                AddAssetParametersToCommand(oAsset)
                _database.ExecuteSP_NO_Return("Asset_Update")
            End Sub

            Private Sub AddAssetParametersToCommand(ByRef oAsset As Asset)
                With _database
                    .AddParameter("@SiteID", oAsset.PropertyID, True)
                    .AddParameter("@ProductID", oAsset.ProductID, True)
                    .AddParameter("@SerialNum", oAsset.SerialNum, True)
                    If oAsset.DateFitted = Date.MinValue Or oAsset.DateFitted = Nothing Then
                        .AddParameter("@DateFitted", DBNull.Value, True)
                    Else
                        .AddParameter("@DateFitted", oAsset.DateFitted, True)
                    End If

                    .AddParameter("@Location", oAsset.Location, True)

                    If oAsset.CertificateLastIssued = Date.MinValue Or oAsset.CertificateLastIssued = Nothing Then
                        .AddParameter("@CertificateLastIssued", DBNull.Value, True)
                    Else
                        .AddParameter("@CertificateLastIssued", oAsset.CertificateLastIssued, True)
                    End If

                    If oAsset.LastServicedDate = Date.MinValue Or oAsset.LastServicedDate = Nothing Then
                        .AddParameter("@LastServicedDate", DBNull.Value, True)
                    Else
                        .AddParameter("@LastServicedDate", oAsset.LastServicedDate, True)
                    End If

                    If oAsset.WarrantyStartDate = Date.MinValue Or oAsset.WarrantyStartDate = Nothing Then
                        .AddParameter("@WarrantyStartDate", DBNull.Value, True)
                    Else
                        .AddParameter("@WarrantyStartDate", oAsset.WarrantyStartDate, True)
                    End If

                    If oAsset.WarrantyEndDate = Date.MinValue Or oAsset.WarrantyEndDate = Nothing Then
                        .AddParameter("@WarrantyEndDate", DBNull.Value, True)
                    Else
                        .AddParameter("@WarrantyEndDate", oAsset.WarrantyEndDate, True)
                    End If

                    .AddParameter("@BoughtFrom", oAsset.BoughtFrom, True)
                    .AddParameter("@SuppliedBy", oAsset.SuppliedBy, True)

                    .AddParameter("@Notes", oAsset.Notes, True)

                    .AddParameter("@GCNumber", oAsset.GCNumber, True)
                    .AddParameter("@YearOfManufacture", oAsset.YearOfManufacture, True)
                    .AddParameter("@ApproximateValue", oAsset.ApproximateValue, True)
                    .AddParameter("@GasTypeID", oAsset.GasTypeID, True)
                    .AddParameter("@FlueTypeID", oAsset.FlueTypeID, True)
                    .AddParameter("@Active", oAsset.Active, True)
                    .AddParameter("@TenantsAppliance", oAsset.TenantsAppliance, True)

                End With
            End Sub

#End Region

        End Class

    End Namespace

End Namespace

File: apollo/apollo.fsm/src/Application Specific/Classes/Entity Collections/Asset/AssetSQL.cs
using System;
using System.Data;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace Assets
    {
        public class AssetSQL
        {
            private Sys.Database _database;

            public AssetSQL(Sys.Database database)
            {
                _database = database;
            }

            public void Delete(int AssetID)
            {
                _database.ClearParameter();
                _database.AddParameter("@AssetID", AssetID, true);
                _database.ExecuteSP_NO_Return("Asset_Delete");
            }

            public Asset Asset_Get(int AssetID)
            {
                _database.ClearParameter();
                _database.AddParameter("@AssetID", AssetID);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("Asset_Get");
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oAsset = new Asset();
                        oAsset.IgnoreExceptionsOnSetMethods = true;
                        oAsset.SetAssetID = dt.Rows[0]["AssetID"];
                        oAsset.SetPropertyID = dt.Rows[0]["SiteID"];
                        oAsset.SetGasTypeID = Sys.Helper.MakeIntegerValid(dt.Rows[0]["GasTypeID"]);
                        oAsset.SetFlueTypeID = Sys.Helper.MakeIntegerValid(dt.Rows[0]["FlueTypeID"]);
                        oAsset.SetProductID = dt.Rows[0]["ProductID"];
                        oAsset.SetSerialNum = Sys.Helper.MakeStringValid(dt.Rows[0]["SerialNum"]);
                        if (Information.IsDBNull(dt.Rows[0]["DateFitted"]))
                        {
                            oAsset.DateFitted = DateTime.MinValue;
                        }
                        else
                        {
                            oAsset.DateFitted = Conversions.ToDate(dt.Rows[0]["DateFitted"]);
                        }

                        if (Information.IsDBNull(dt.Rows[0]["CertificateLastIssued"]))
                        {
                            oAsset.CertificateLastIssued = DateTime.MinValue;
                        }
                        else
                        {
                            oAsset.CertificateLastIssued = Conversions.ToDate(dt.Rows[0]["CertificateLastIssued"]);
                        }

                        if (Information.IsDBNull(dt.Rows[0]["LastServicedDate"]))
                        {
                            oAsset.LastServicedDate = DateTime.MinValue;
                        }
                        else
                        {
                            oAsset.LastServicedDate = Conversions.ToDate(dt.Rows[0]["LastServicedDate"]);
                        }

                        if (Information.IsDBNull(dt.Rows[0]["WarrantyStartDate"]))
                        {
                            oAsset.WarrantyStartDate = DateTime.MinValue;
                        }
                        else
                        {
                            oAsset.WarrantyStartDate = Conversions.ToDate(dt.Rows[0]["WarrantyStartDate"]);
                        }

                        if (Information.IsDBNull(dt.Rows[0]["WarrantyEndDate"]))
                        {
                            oAsset.WarrantyEndDate = DateTime.MinValue;
                        }
                        else
                        {
                            oAsset.WarrantyEndDate = Conversions.ToDate(dt.Rows[0]["WarrantyEndDate"]);
                        }

                        oAsset.SetBoughtFrom = Sys.Helper.MakeStringValid(dt.Rows[0]["BoughtFrom"]);
                        oAsset.SetSuppliedBy = Sys.Helper.MakeStringValid(dt.Rows[0]["SuppliedBy"]);
                        oAsset.SetLocation = Sys.Helper.MakeStringValid(dt.Rows[0]["Location"]);
                        oAsset.SetNotes = Sys.Helper.MakeStringValid(dt.Rows[0]["Notes"]);
                        oAsset.SetGCNumber = Sys.Helper.MakeStringValid(dt.Rows[0]["GCNumber"]);
                        oAsset.SetYearOfManufacture = Sys.Helper.MakeStringValid(dt.Rows[0]["YearOfManufacture"]);
                        oAsset.SetApproximateValue = Sys.Helper.MakeDoubleValid(dt.Rows[0]["ApproximateValue"]);
                        oAsset.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oAsset.SetActive = Sys.Helper.MakeBooleanValid(dt.Rows[0]["Active"]);
                        oAsset.SetTenantsAppliance = Sys.Helper.MakeBooleanValid(dt.Rows[0]["TenantsAppliance"]);
                        oAsset.Exists = true;
                        return oAsset;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public DataView Asset_GetAll(int userID)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", userID);
                var dt = _database.ExecuteSP_DataTable("Asset_GetAll_Mk1");
                dt.TableName = Sys.Enums.TableNames.tblAsset.ToString();
                return new DataView(dt);
            }

            public DataView Asset_GetForSite(int SiteID)
            {
                _database.ClearParameter();
                _database.AddParameter("@SiteID", SiteID, true);
                var dt = _database.ExecuteSP_DataTable("[Asset_GetForSite]");
                dt.TableName = Sys.Enums.TableNames.tblAsset.ToString();
                return new DataView(dt);
            }

            public DataView Asset_Search(string Criteria, int RegionID = 0)
            {
                _database.ClearParameter();
                _database.AddParameter("@SearchString", Criteria, true);
                _database.AddParameter("@RegionID", RegionID, true);
                var dt = _database.ExecuteSP_DataTable("[Asset_SearchList]");
                dt.TableName = Sys.Enums.TableNames.tblAsset.ToString();
                return new DataView(dt);
            }

            public Asset Insert(Asset oAsset)
            {
                _database.ClearParameter();
                AddAssetParametersToCommand(ref oAsset);
                oAsset.SetAssetID = Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Asset_Insert"));
                oAsset.Exists = true;
                return oAsset;
            }

            public void Update(Asset oAsset)
            {
                _database.ClearParameter();
                _database.AddParameter("@AssetID", oAsset.AssetID, true);
                AddAssetParametersToCommand(ref oAsset);
                _database.ExecuteSP_NO_Return("Asset_Update");
            }

            private void AddAssetParametersToCommand(ref Asset oAsset)
            {
                {
                    var withBlock = _database;
                    withBlock.AddParameter("@SiteID", oAsset.PropertyID, true);
                    withBlock.AddParameter("@ProductID", oAsset.ProductID, true);
                    withBlock.AddParameter("@SerialNum", oAsset.SerialNum, true);
                    if (oAsset.DateFitted == DateTime.MinValue | oAsset.DateFitted == default)
                    {
                        withBlock.AddParameter("@DateFitted", DBNull.Value, true);
                    }
                    else
                    {
                        withBlock.AddParameter("@DateFitted", oAsset.DateFitted, true);
                    }

                    withBlock.AddParameter("@Location", oAsset.Location, true);
                    if (oAsset.CertificateLastIssued == DateTime.MinValue | oAsset.CertificateLastIssued == default)
                    {
                        withBlock.AddParameter("@CertificateLastIssued", DBNull.Value, true);
                    }
                    else
                    {
                        withBlock.AddParameter("@CertificateLastIssued", oAsset.CertificateLastIssued, true);
                    }

                    if (oAsset.LastServicedDate == DateTime.MinValue | oAsset.LastServicedDate == default)
                    {
                        withBlock.AddParameter("@LastServicedDate", DBNull.Value, true);
                    }
                    else
                    {
                        withBlock.AddParameter("@LastServicedDate", oAsset.LastServicedDate, true);
                    }

                    if (oAsset.WarrantyStartDate == DateTime.MinValue | oAsset.WarrantyStartDate == default)
                    {
                        withBlock.AddParameter("@WarrantyStartDate", DBNull.Value, true);
                    }
                    else
                    {
                        withBlock.AddParameter("@WarrantyStartDate", oAsset.WarrantyStartDate, true);
                    }

                    if (oAsset.WarrantyEndDate == DateTime.MinValue | oAsset.WarrantyEndDate == default)
                    {
                        withBlock.AddParameter("@WarrantyEndDate", DBNull.Value, true);
                    }
                    else
                    {
                        withBlock.AddParameter("@WarrantyEndDate", oAsset.WarrantyEndDate, true);
                    }

                    withBlock.AddParameter("@BoughtFrom", oAsset.BoughtFrom, true);
                    withBlock.AddParameter("@SuppliedBy", oAsset.SuppliedBy, true);
                    withBlock.AddParameter("@Notes", oAsset.Notes, true);
                    withBlock.AddParameter("@GCNumber", oAsset.GCNumber, true);
                    withBlock.AddParameter("@YearOfManufacture", oAsset.YearOfManufacture, true);
                    withBlock.AddParameter("@ApproximateValue", oAsset.ApproximateValue, true);
                    withBlock.AddParameter("@GasTypeID", oAsset.GasTypeID, true);
                    withBlock.AddParameter("@FlueTypeID", oAsset.FlueTypeID, true);
                    withBlock.AddParameter("@Active", oAsset.Active, true);
                    withBlock.AddParameter("@TenantsAppliance", oAsset.TenantsAppliance, true);
                }
            }
        }
    }
}